google/gax/__init__.py

```diff
@@ -429,3 +431,33 @@ def next(self):
         if not self.page_token:
             self._done = True
         return getattr(response, self._page_descriptor.resource_field)
+
+
+class ResourceIterator(object):
+    """An iterator over resources of the page iterator."""
+
+    # pylint: disable=too-few-public-methods
+    def __init__(self, page_iterator):
+        """Constructor.
+
+        Args:
+          page_iterator (PageIterator): the base iterator of getting pages.
+        """
+        self._iterator = page_iterator
+        self._current = None
+        self._index = -1
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        """Retrieves the next resource."""
+        # pylint: disable=next-method-called
+        while not self._current:
```

Comment:
PageIterator could return an empty array but have a next page token, in that case this should skip the empty array and fetch the next page.